File: frontend/src/services/matchService.ts
import { apiClient } from './api'

/**
 * Match Service
 * Handles all match-related API calls
 */

export interface Match {
  id: number
  home_team_id: number
  away_team_id: number
  home_team_name?: string
  away_team_name?: string
  league: string
  season: string
  match_date: string
  venue: string
  home_score: number | null
  away_score: number | null
  status: 'scheduled' | 'ongoing' | 'completed' | 'postponed'
  possession_home: number | null
  possession_away: number | null
  shots_home: number | null
  shots_away: number | null
  created_at: string
  updated_at: string
}

export interface MatchStats {
  match_id: number
  possession_home: number
  possession_away: number
  shots: number
  shots_on_target: number
  passes: number
  tackles: number
  fouls: number
}

export interface MatchResponse {
  status: string
  data: Match | Match[]
  message: string
}

class MatchService {
  /**
   * Get all matches
   */
  static async getMatches(league?: string, status = 'all', limit = 20) {
    try {
      const response = await apiClient.get<MatchResponse>('/matches', {
        params: { league, status, limit },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching matches:', error)
      throw error
    }
  }

  /**
   * Get match by ID
   */
  static async getMatchById(matchId: number) {
    try {
      const response = await apiClient.get<MatchResponse>(`/matches/${matchId}`)
      return response.data
    } catch (error) {
      console.error(`Error fetching match ${matchId}:`, error)
      throw error
    }
  }

  /**
   * Get upcoming matches
   */
  static async getUpcomingMatches(limit = 10, league?: string) {
    try {
      const response = await apiClient.get<MatchResponse>('/matches/upcoming', {
        params: { limit, league },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching upcoming matches:', error)
      throw error
    }
  }

  /**
   * Get recent matches
   */
  static async getRecentMatches(limit = 10) {
    try {
      const response = await apiClient.get<MatchResponse>('/matches/recent', {
        params: { limit },
      })
      return response.data
    } catch (error) {
      console.error('Error fetching recent matches:', error)
      throw error
    }
  }

  /**
   * Get team matches
   */
  static async getTeamMatches(teamId: number, limit = 20) {
    try {
      const response = await apiClient.get<MatchResponse>(`/teams/${teamId}/matches`, {
        params: { limit },
      })
      return response.data
    } catch (error) {
      console.error(`Error fetching matches for team ${teamId}:`, error)
      throw error
    }
  }

  /**
   * Get head-to-head matches
   */
  static async getHeadToHeadMatches(teamId1: number, teamId2: number) {
    try {
      const response = await apiClient.get<MatchResponse>(`/matches/${teamId1}/head-to-head`, {
        params: { opponent_id: teamId2 },
      })
      return response.data
    } catch (error) {
      console.error(`Error fetching h2h matches:`, error)
      throw error
    }
  }

  /**
   * Get match statistics
   */
  static async getMatchStats(matchId: number) {
    try {
      const response = await apiClient.get<MatchResponse>(`/matches/${matchId}/stats`)
      return response.data
    } catch (error) {
      console.error(`Error fetching stats for match ${matchId}:`, error)
      throw error
    }
  }

  /**
   * Get match events
   */
  static async getMatchEvents(matchId: number) {
    try {
      const response = await apiClient.get<MatchResponse>(`/matches/${matchId}/events`)
      return response.data
    } catch (error) {
      console.error(`Error fetching events for match ${matchId}:`, error)
      throw error
    }
  }

  /**
   * Get player match performance
   */
  static async getPlayerMatchPerformance(matchId: number, playerId: number) {
    try {
      const response = await apiClient.get<MatchResponse>(
        `/matches/${matchId}/players/${playerId}`
      )
      return response.data
    } catch (error) {
      console.error(
        `Error fetching player ${playerId} performance in match ${matchId}:`,
        error
      )
      throw error
    }
  }
}

export default MatchService

File: backend/services/match_service.py
"""Match Service - Business logic for match operations"""

from typing import List, Optional, Dict
import logging
from datetime import datetime

logger = logging.getLogger(__name__)


class MatchService:
    """Service class for match-related operations"""
    
    def __init__(self):
        """Initialize match service"""
        logger.info("✅ MatchService initialized")
    
    @staticmethod
    def get_all_matches(
        league: Optional[str] = None,
        status: str = 'all',
        limit: int = 20
    ) -> List[Dict]:
        """Get matches with filters"""
        
        logger.info(f"Getting matches - League: {league}, Status: {status}")
        
        # TODO: Query matches from database
        
        return []
    
    @staticmethod
    def get_match_by_id(match_id: int) -> Optional[Dict]:
        """Get specific match details"""
        
        logger.info(f"Getting match: {match_id}")
        
        # TODO: Query database
        
        match = {
            'id': match_id,
            'home_team': 'Team A',
            'away_team': 'Team B',
            'home_score': 2,
            'away_score': 1,
            'date': '2025-05-25T15:00:00',
            'status': 'completed',
            'venue': 'Stadium Name',
            'league': 'League Name'
        }
        
        return match
    
    @staticmethod
    def get_upcoming_matches(limit: int = 10, league: Optional[str] = None) -> List[Dict]:
        """Get upcoming matches"""
        
        logger.info(f"Getting {limit} upcoming matches")
        
        # TODO: Query future matches
        
        return []
    
    @staticmethod
    def get_recent_matches(limit: int = 10) -> List[Dict]:
        """Get recently completed matches"""
        
        logger.info(f"Getting {limit} recent matches")
        
        # TODO: Query past matches
        
        return []
    
    @staticmethod
    def get_team_recent_matches(team_id: int, limit: int = 10) -> List[Dict]:
        """Get team's recent matches"""
        
        logger.info(f"Getting {limit} recent matches for team: {team_id}")
        
        # TODO: Filter matches by team
        
        return []
    
    @staticmethod
    def get_head_to_head_matches(team_id_1: int, team_id_2: int) -> List[Dict]:
        """Get all historical matches between two teams"""
        
        logger.info(f"Getting h2h matches: Team {team_id_1} vs Team {team_id_2}")
        
        # TODO: Query historical matches
        
        return []
    
    @staticmethod
    def get_match_statistics(match_id: int) -> Dict:
        """Get detailed match statistics"""
        
        logger.info(f"Getting statistics for match: {match_id}")
        
        # TODO: Calculate possession, shots, passes, etc.
        
        stats = {
            'match_id': match_id,
            'possession': {
                'home': 52,
                'away': 48
            },
            'shots': {
                'home': 15,
                'away': 10
            },
            'shots_on_target': {
                'home': 5,
                'away': 3
            },
            'passes': {
                'home': 600,
                'away': 550
            },
            'fouls': {
                'home': 12,
                'away': 15
            }
        }
        
        return stats
    
    @staticmethod
    def get_match_events(match_id: int) -> List[Dict]:
        """Get match events (goals, cards, substitutions, etc.)"""
        
        logger.info(f"Getting events for match: {match_id}")
        
        # TODO: Query match events
        
        events = []
        
        return events
    
    @staticmethod
    def get_player_match_performance(match_id: int, player_id: int) -> Dict:
        """Get player's performance in specific match"""
        
        logger.info(f"Getting performance for player {player_id} in match {match_id}")
        
        # TODO: Query player stats from match
        
        performance = {
            'player_id': player_id,
            'match_id': match_id,
            'goals': 0,
            'assists': 0,
            'rating': 7.5,
            'minutes_played': 90,
            'passes': 50,
            'pass_accuracy': 85.0
        }
        
        return performance
    
    @staticmethod
    def get_team_performance_in_match(match_id: int, team_id: int) -> Dict:
        """Get team's performance in specific match"""
        
        logger.info(f"Getting team {team_id} performance in match {match_id}")
        
        # TODO: Calculate team stats for match
        
        performance = {
            'team_id': team_id,
            'match_id': match_id,
            'goals': 2,
            'shots': 15,
            'possession': 52
        }
        
        return performance
    
    @staticmethod
    def get_matches_by_league(league: str, limit: int = 20) -> List[Dict]:
        """Get matches from specific league"""
        
        logger.info(f"Getting {limit} matches from {league}")
        
        # TODO: Filter by league
        
        return []
    
    @staticmethod
    def get_matches_by_date(date: str) -> List[Dict]:
        """Get all matches on specific date"""
        
        logger.info(f"Getting matches on {date}")
        
        # TODO: Filter by date
        
        return []
